File: frontend/src/api/progress.ts
import axios from 'axios';
import { API_BASE_URL } from '../config/env';
import { submitReliably } from './submitQueue';

// 创建axios实例
const apiClient = axios.create({
  baseURL: API_BASE_URL,
});

// 配置axios拦截器,自动添加token
apiClient.interceptors.request.use(
  (config) => {
    const token = localStorage.getItem('access_token');
    if (token) {
      config.headers.Authorization = `Bearer ${token}`;
    }
    return config;
  },
  (error) => {
    return Promise.reject(error);
  }
);

// 响应拦截器 - 统一处理错误
apiClient.interceptors.response.use(
  (response) => response,
  (error) => {
    // 提取友好的错误信息
    if (error.response?.data?.detail) {
      // 创建新的错误对象,包含后端返回的详细信息
      const friendlyError = new Error(error.response.data.detail);
      friendlyError.name = error.name;
      return Promise.reject(friendlyError);
    }
    return Promise.reject(error);
  }
);

// ========================================
// 学生端学习进度API
// ========================================


export interface StartLearningRequest {
  unit_id: number;
  learning_mode: string;
}

export interface WordData {
  id: number;
  word: string;
  phonetic: string | null;
  syllables: string | null;
  tts_text: string | null;
  difficulty: number;
  audio_url: string | null;
  image_url: string | null;
  order_index: number;
  meaning: string | null;
  part_of_speech: string | null;
  example_sentence: string | null;
  example_translation: string | null;
}

export interface StartLearningResponse {
  has_existing_progress: boolean;
  current_word_index: number;
  completed_words: number;
  total_words: number;
  progress_percentage: number;
  words: WordData[];
  message: string;
  unit_info: {
    id: number;
    unit_number: number;
    name: string;
    description: string | null;
    book_id: number;
    grade_level: string | null;
    group_size?: number;
  };
}

export interface UpdateProgressRequest {
  unit_id: number;
  learning_mode: string;
  current_word_index: number;
  current_word_id?: number;
  word_result?: string;
  is_completed: boolean;
}

export interface UpdateProgressResponse {
  success: boolean;
  message: string;
  progress_percentage: number;
  completed_words: number;
  total_words: number;
  is_completed: boolean;
}

export interface UnitProgress {
  unit_id: number;
  unit_number: number;
  unit_name: string;
  word_count: number;
  completed_words: number;
  progress_percentage: number;
  has_progress: boolean;
  current_word_index: number;
  last_studied_at: string | null;
  learning_mode: string | null;
  is_completed: boolean;
  best_accuracy: number | null;
  is_perfect: boolean;
  total_study_time?: number;
  attempt_count?: number;
  // 严格模式:false = 未在教师分配范围内,锁定不可学
  is_allowed?: boolean;
}

export interface BookProgress {
  book_id: number;
  book_name: string;
  grade_level: string | null;
  volume: string | null;
  unit_count: number;
  word_count: number;
  completed_words: number;
  progress_percentage: number;
  units: UnitProgress[];
}

export interface StudentBook {
  id: number;
  name: string;
  description: string | null;
  grade_level: string | null;
  volume: string | null;
  cover_color: string;
  cover_url: string | null;
  unit_count: number;
  word_count: number;
  progress_percentage: number;
  owned: boolean;
  created_at: string;
}

// API函数

export const startLearning = async (request: StartLearningRequest): Promise<StartLearningResponse> => {
  const response = await apiClient.post(
    `/student/units/${request.unit_id}/start`,
    { learning_mode: request.learning_mode }  // 只发送learning_mode,unit_id已在URL中
  );
  return response.data;
};

// 进度更新是绝对值写入(天然幂等),走可靠队列:失败自动补交,
// 最后一组的 is_completed 丢了会导致单元不解锁/轮数不涨,必须保证送达
export const updateProgress = async (request: UpdateProgressRequest): Promise<UpdateProgressResponse> => {
  return submitReliably('/student/progress', request, {
    method: 'put',
    dedupe: false,
    staleKey: `progress:${request.unit_id}:${request.learning_mode}`,
  });
};

export const getBookProgress = async (bookId: number): Promise<BookProgress> => {
  const response = await apiClient.get(`/student/books/${bookId}/progress`);
  return response.data;
};

export const getUnitProgress = async (unitId: number, learningMode: string): Promise<UnitProgress> => {
  const response = await apiClient.get(
    `/student/units/${unitId}/progress?learning_mode=${learningMode}`
  );
  return response.data;
};

export const getStudentBooks = async (): Promise<StudentBook[]> => {
  const response = await apiClient.get(`/student/books`);
  return response.data;
};

// 类型已在文件顶部使用export interface导出

File: backend/app/schemas/progress.py
from pydantic import BaseModel, Field
from typing import List, Optional
from datetime import datetime

# ========================================
# 学习进度相关 Schemas
# ========================================

class LearningProgressBase(BaseModel):
    """学习进度基础模型"""
    unit_id: int = Field(..., description="单元ID")
    learning_mode: str = Field(..., description="学习模式: flashcard/quiz/spelling/fillblank")

class StartLearningRequest(BaseModel):
    """开始学习请求 - unit_id从路径参数获取"""
    learning_mode: str = Field(..., description="学习模式: flashcard/quiz/spelling/fillblank")

class StartLearningResponse(BaseModel):
    """开始学习响应"""
    has_existing_progress: bool = Field(..., description="是否有已存在的学习进度")
    current_word_index: int = Field(..., description="当前单词索引(从0开始)")
    completed_words: int = Field(..., description="已完成单词数")
    total_words: int = Field(..., description="总单词数")
    progress_percentage: float = Field(..., description="进度百分比")
    words: List[dict] = Field(..., description="单词列表")
    message: str = Field(..., description="提示信息")
    unit_info: dict = Field(..., description="单元信息")

class UpdateProgressRequest(BaseModel):
    """更新学习进度请求"""
    unit_id: int = Field(..., description="单元ID")
    learning_mode: str = Field(..., description="学习模式")
    current_word_index: int = Field(..., description="当前单词索引")
    current_word_id: Optional[int] = Field(None, description="当前单词ID")
    word_result: Optional[str] = Field(None, description="单词学习结果: know/dont_know/correct/wrong")
    is_completed: bool = Field(False, description="是否完成该单元学习")

class UpdateProgressResponse(BaseModel):
    """更新学习进度响应"""
    success: bool = Field(..., description="是否成功")
    message: str = Field(..., description="提示信息")
    progress_percentage: float = Field(..., description="进度百分比")
    completed_words: int = Field(..., description="已完成单词数")
    total_words: int = Field(..., description="总单词数")
    is_completed: bool = Field(..., description="是否完成")

# ========================================
# 单元进度相关
# ========================================

class UnitProgressResponse(BaseModel):
    """单元进度响应"""
    unit_id: int
    unit_number: int
    unit_name: str
    word_count: int
    completed_words: int = Field(0, description="已完成单词数")
    progress_percentage: float = Field(0.0, description="进度百分比")
    has_progress: bool = Field(False, description="是否有学习进度")
    current_word_index: int = Field(0, description="当前单词索引")
    last_studied_at: Optional[datetime] = Field(None, description="最后学习时间")
    learning_mode: Optional[str] = Field(None, description="学习模式")
    is_completed: bool = Field(False, description="是否完成")
    best_accuracy: Optional[float] = Field(None, description="最佳正确率")
    is_perfect: bool = Field(False, description="是否满分通过过")
    total_study_time: int = Field(0, description="总学习时长(秒)")
    attempt_count: int = Field(0, description="学习轮次(会话数)")

class BookProgressResponse(BaseModel):
    """单词本进度响应"""
    book_id: int
    book_name: str
    unit_count: int = Field(..., description="单元数量")
    word_count: int = Field(..., description="总单词数")
    completed_words: int = Field(0, description="已完成单词数")
    progress_percentage: float = Field(0.0, description="整体进度百分比")
    units: List[UnitProgressResponse] = Field(default_factory=list, description="单元进度列表")

# ========================================
# 学生单词本列表
# ========================================

class StudentBookListItem(BaseModel):
    """学生单词本列表项"""
    id: int
    name: str
    description: Optional[str]
    grade_level: Optional[str]
    volume: Optional[str] = None
    cover_color: str
    unit_count: int = Field(0, description="单元数量")
    word_count: int = Field(0, description="总单词数")
    progress_percentage: float = Field(0.0, description="学习进度百分比")
    owned: bool = Field(False, description="是否已购买/已分配")
    created_at: datetime

    class Config:
        from_attributes = True
